src/models/assistant-model.ts

```diff
@@ -86,60 +101,104 @@ export const AssistantModel = types
+         content: formatJsonMessage(runState.status),
+       });
+     }
+
+     if (errorStates.includes(runState.status)) {
+      self.transcriptStore.addMessage(DEBUG_SPEAKER, {
+         description: "Run failed",
+         content: formatJsonMessage(runState),
+       });
+       self.transcriptStore.addMessage(DAVAI_SPEAKER, {
+         content: "I'm sorry, I encountered an error. Please try again.",
+       });
+     }
+
+     if (runState.status === "requires_action") {
+      self.transcriptStore.addMessage(DEBUG_SPEAKER, {
+         description: "Run requires action",
+         content: formatJsonMessage(runState),
+       });
+       yield handleRequiredAction(runState, currentRunId);
+       yield pollRunState(currentRunId);
+     }
+
+     if (runState.status === "completed") {
+       const messages = yield self.apiConnection.beta.threads.messages.list(self.thread.id);
+
+       const lastMessageForRun = messages.data
+         .filter((msg: Message) => msg.run_id === currentRunId && msg.role === "assistant")
+         .pop();
+
+         self.transcriptStore.addMessage(DEBUG_SPEAKER, {
+         description: "Run completed, assistant response",
+         content: formatJsonMessage(lastMessageForRun),
+       });
+
+       const lastMessageContent = lastMessageForRun?.content[0]?.text?.value;
+       if (lastMessageContent) {
+        self.transcriptStore.addMessage(DAVAI_SPEAKER, { content: lastMessageContent });
+       } else {
+        self.transcriptStore.addMessage(DAVAI_SPEAKER, {
+           content: "I'm sorry, I don't have a response for that.",
+         });
+       }
+     }
+   });
+
     const handleRequiredAction = flow(function* (runState, runId) {
       try {
         const toolOutputs = runState.required_action?.submit_tool_outputs.tool_calls
           ? yield Promise.all(
             runState.required_action.submit_tool_outputs.tool_calls.map(flow(function* (toolCall: any) {
-              if (toolCall.function.name === "get_attributes") {
-                const { dataset } = JSON.parse(toolCall.function.arguments);
-                // getting the root collection won't always work. what if a user wants the attributes
-                // in the Mammals dataset but there is a hierarchy?
-                const rootCollection = (yield getDataContext(dataset)).values.collections[0];
-                const attributeList = yield getAttributeList(dataset, rootCollection.name);
-                return { tool_call_id: toolCall.id, output: JSON.stringify(attributeList) };
+              if (toolCall.function.name === "create_request") {
+                const { action, resource, values } = JSON.parse(toolCall.function.arguments);
+                const request = { action, resource, values };
+                self.transcriptStore.addMessage(DEBUG_SPEAKER, { description: "Request sent to CODAP", content: formatJsonMessage(request) });
+                const res = yield codapInterface.sendRequest(request);
```

Comment:
This may be a later addition, but it seems like some sort of sanity checking before sending it direct to CODAP might be a good idea?  Are there API calls that would cause real problems, or crash the page?

Reply:
I don't think there are any that will crash the page... I think at worst you would just get a { success: false } message.

In theory it could do something weird like create a request to delete something you don't want to delete. But, if we implement the undo/redo button functionality, I guess that would be a reversible scenario.

I think this will be something to keep in mind moving forward as we test it out with different inputs.